ssl/s3_lib.c

```diff
@@ -3466,6 +3466,25 @@ long ssl3_ctrl(SSL *s, int cmd, long larg, void *parg)
         break;
 #endif                          /* !OPENSSL_NO_EC */
     case SSL_CTRL_SET_TLSEXT_HOSTNAME:
+#if 0
+        /*
+         * TODO(OpenSSL1.2) s->server is initially set based on the absence
+         * of an accept method, and can be a false positive when a generic
+         * SSL_METHOD is used.  We rely on the ability to set a (client)
+         * SNI hostname before forcing 'client' status, as tested in
+         * test_servername.c:client_setup_sni_before_state(), so this
+         * block of code cannot be enabled until ABI-breaking changes
+         * are permitted by release policy.
+         */
+        /*
+         * This is the API to set what SNI a client requests; it makes no
+         * sense for a server.
+         */
+        if (s->server) {
+            SSLerr(SSL_F_SSL3_CTRL, SSL_R_SSL3_EXT_INVALID_SERVERNAME);
+            return 0;
```

Comment:
Perhaps return success for this condition, but do nothing? The unit tests will then pass (or change the unit tests, but then this breaks existing behavior...)

Reply:
The OpenSSL API is such that `s->server` is not meaningful in config functions. So you, in general, need to make your config functions always stash the value, but just ignore it if not applicable. E.g.

```
  SSL_CTX *ctx = SSL_CTX_new(TLS_method());
  SSL *ssl = SSL_new(ctx);
  SSL_set_something(ssl);
  SSL_set_something_else(ssl);
  // ssl->server is not set until this point:
  SSL_connect(ssl);  // Or SSL_accept(ssl)
```

Reply:
> The OpenSSL API is such that ```s->server``` is not meaningful in config functions.

The current one is, yes.  Perhaps it need not be, though (when we are permitted to make such breaking changes).  But I don't think there would be too much harm from just dropping the last commit, if I'm in the minority.  As a very far aside note, I've been pondering whether there is room to do a full-on API modernization for 1.2, making a new TLS structure and TLS_foo() public APIs that are free of historical cruft (such as this lack of meaning for ```s->server``` until the handshake begins).

I could also do as you implicitly suggest, and defer this check until SSL_do_handshake(), if people want that, of course.

Reply:
Bifurcating all the APIs is kind of expensive, as folks tend to support a couple of OpenSSL versions at the time. At the least, before doing that, *definitely* provide a compatibility story. 1.1.0 is not a good model.

But you could fix this more simply. Just get rid of `TLS_method` and say that you want people to use `TLS_client_method` and `TLS_server_method`. I considered that for BoringSSL, but found that too much existing code used what was then `SSLv23_method` to desupport the uncommitted state. But with a proper deprecation path you could do it.

Though having the server setters "work" for clients and vice versa isn't the end of the world... except where OpenSSL has those problematic combined config and state getters. You've identified one of them in this PR. Those have time and time again demonstrated themselves to be an hindrance for improvements on our end. (We organize state not by whether it happens to be an extension, but by scope, which lets us easily, e.g., release all temporary handshake state after the handshake completes. Or cleanly support `SSL_clear`. But by clearly separating that state, it becomes very obvious each time some corresponding getter is ambiguous on what it actually does.)

Reply:
I'm not sure I like the idea of adding `if 0` code here. It tends to rot. Just a simple comment would probably be better.

Reply:
To be honest, I wasn't even sure if any of the last commit would stick around.  But I can change it to just a comment.